Add developer dependencies into vendor/

aanm opened the pull request:
Fixes: https://github.com/cilium/cilium/issues/2959

I may have overdone it for some dependencies, if you don't agree I can revert the commits for each one.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/4353)
<!-- Reviewable:end -->

aanm committed:
vendor: remove go-bindata dependency from the developer requirements

Since dep allows to add dependencies for generators without importing
them in the project we can stop requiring the developers to install the
go-bindata dependency locally.

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
vendor: remove ginkgo dependency from the developer requirements

Since dep allows to add dependencies for generators without importing
them in the project we can stop requiring the developers to install the
ginkgo dependency locally.

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
vendor: remove gomega dependency from the developer requirements

Since dep allows to add dependencies for generators without importing
them in the project we can stop requiring the developers to install the
gomega dependency locally.

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
vendor: add protoc-gen-go to install dependencies target

Add protoc-gen-go to install dependencies target to make it easier
for new contributors to install the necessary requirements to build Cilium.

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
vendor: add protoc-gen-validate to install dependencies target

Add protoc-gen-validate to install dependencies target to make it easier
for new contributors to install the necessary requirements to build Cilium.

Signed-off-by: André Martins <andre@cilium.io>

ianvernon commented on Makefile:
```diff
@@ -262,6 +262,9 @@ test-docs: docs-container
 	docker container run --rm -v $$(pwd):/srv/ cilium/docs-builder /bin/bash -c 'make html'
 
+install-cilium-golang-deps:
+	curl -SsL https://github.com/ksonnet/kubecfg/releases/download/v0.8.0/kubecfg-linux-amd64 > ../../../../bin/kubecfg && chmod +x ../../../../bin/kubecfg
```
What exactly is the location of this `bin/kubecfg` directory? Why not use `GOPATH`?

aanm replied:
@ianvernon 

GOPATH is not required to be present in a golang environment

that bin directory will be `${GOPATH}/bin`

ianvernon replied:
Didn't know that, TIL. LGTM then apart from the build failure.

raybejjani commented on daemon/Makefile:
```diff
@@ -13,7 +13,9 @@ $(TARGET): $(SOURCES) check-bindata
 	$(GO) build -i $(GOBUILD) -o $(TARGET)
 
-GO_BINDATA := $(QUIET) go-bindata -prefix ../ -mode 0640 -modtime 1450269211 \
+GO_BINDATA_BIN := $(GO) run ../vendor/github.com/cilium/go-bindata/go-bindata/*.go
```
Could we just go install go-bindata? it's a little weird to me to invoke it so directly. The path includes the correct bin directory, right?

aanm replied:
it's `$(GO) run` compiles the program at runtime.

I don't like it either but I made this commit before adding the target `install-cilium-golang-deps`

raybejjani replied:
Assuming you agree, we could use the `-o` option to go build to place the tool binaries in a local `bin` directory, then modify path to prefer this `bin`. This also seems bad, though, since I usually expect make to use the PATH I give it...

All of this is to avoid asking the user to install the tools, right? I guess it also ensures the correct version is used.

raybejjani commented on Documentation/contributing.rst:
```diff
@@ -33,8 +33,6 @@ contribute to Cilium:
 | `dep <https://github.com/golang/dep/>`_                                          | >= v0.4.1                | ``curl https://raw.githubusercontent.com/golang/dep/master/install.sh | sh``  |
 +----------------------------------------------------------------------------------+--------------------------+-------------------------------------------------------------------------------+
-| `go-bindata <https://github.com/cilium/go-bindata>`_                             | ``a0ff2567cfb``          | ``go get -u github.com/cilium/go-bindata/...``                                |
-+----------------------------------------------------------------------------------+--------------------------+-------------------------------------------------------------------------------+
```
Unless the goal is to hide the fact that we're using go-bindata it is still required. This list isn't just what to install, it's a quick way to get a sense for the bigger external dependencies. On the other hand, the .lock file does that more comprehensively so I guess it's ok not to mention it iff we continue to use go run in the makefile.

raybejjani commented on Documentation/contributing.rst:
```diff
@@ -37,19 +37,25 @@ contribute to Cilium:
 +----------------------------------------------------------------------------------+--------------------------+-------------------------------------------------------------------------------+
-+ `Kubecfg <https://github.com/ksonnet/kubecfg>`_                                  | >=0.8.0                  | go get github.com/ksonnet/kubecfg                                             |
+
+The following Golang dependencies can be installed by running `make install-cilium-golang-deps`:
```
We should explain when this is used, and so when to install this. Alternatively, you can make the rule `install-cilium-test-deps` (this isn't a go dependency, it's used as a tool).

raybejjani commented on Makefile:
```diff
@@ -264,6 +264,7 @@ test-docs: docs-container
 
 install-cilium-golang-deps:
 	curl -SsL https://github.com/ksonnet/kubecfg/releases/download/v0.8.0/kubecfg-linux-amd64 > ../../../../bin/kubecfg && chmod +x ../../../../bin/kubecfg
+	go build -o ../../../../bin/protoc-gen-go vendor/github.com/golang/protobuf/protoc-gen-go/main.go
```
Wouldn't a go install put this in the correct place? or did I miss a cilium specific bin directory that we're now using in PATH?

aanm replied:
I can't run `go install` perhaps because it's in a vendor directory :thinking:

raybejjani replied:
yeah, it's likely that vendor is only used to resolve dependencies during a build... `go run` is probably doing the right thing (Assuming it's running a full build).

raybejjani reviewed (changes requested):
My main blocker is that I would prefer that we keep the list of dependencies, and list them as included in vendor or something (in a separate section). While that list isn't comprehensive it is the list of what we think are major dependencies. The .lock file is the most correct but no human will read that. If I'm auditing the project quickly this is the list I would want to get a sense for how complex this audit will become.

I'm not sure how I feel about invoking things directly with go run. It's quite brilliant on one hand, but completely unexpected on the other (I'm surprised it works, actually! I didn't think run did all the things go build did like generate templated files etc.). The problem is that we can't simply go install these because that would clobber other installed binaries in gopath :/ The thing is, using go run makes it harder for someone to debug the make rules if it comes up since the invocation is weird looking (but it's already a little obscured but all the variables). In general, I think I still prefer having a install tools make target that we use when building the dev vm.

I don't think anything here could use it, but go generate is one way to avoid some wierdo invocations (https://blog.golang.org/generate). The only candidate might be the bindata files, however.

aanm committed:
dep: avoid dep panic for fsnotify repository

This fix avoids dep 0.4.1 from panic as referenced in
https://github.com/golang/dep/issues/1799#issuecomment-381190029

Signed-off-by: André Martins <andre@cilium.io>
